### PR TITLE
Feature/890 move reservations callbacks and handlers to new functional block

### PR DESCRIPTION
**Pull request by maaikez**
## Describe your changes
Create reservations functional block and move callbacks there. Create unit tests.

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: tests/lib/ocpp/v201/functional_blocks/test_reservation.cpp [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+
+TEST_F(ReservationTest, handle_reserve_now_specific_evse_successful) {
+    // Try to make a reservation for an existing evse, which is accepted.
+    std::optional<ConnectorEnum> tesla_connector_type = ConnectorEnum::cTesla;
+
+    const ocpp::EnhancedMessage<MessageType> request = create_example_reserve_now_request(2, ConnectorEnum::cTesla);
+
+    EvseMock& m1 = evse_manager.get_mock(1);
+    EXPECT_CALL(m1, does_connector_exist(ConnectorEnum::cTesla)).Times(0);
+    EXPECT_CALL(m1, get_connector_status(tesla_connector_type)).Times(0);
```
Comment (marcemmers) [inline tests/lib/ocpp/v201/functional_blocks/test_reservation.cpp:399]: Same here, we don't really care if the code calls this right? Would be weird if it does but it should not fail this as long as the wanted behavior is achieved.

Maybe there is another way to specifically test that it does select the right evse?

**Review by marcemmers — CHANGES_REQUESTED** [review]
Looks good overall. I like that the unit tests are small and to the point, have a few small suggestions there as well as some other small stuff.

File: include/ocpp/v201/functional_blocks/reservation.hpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Pionix GmbH and Contributors to EVerest
+
+#include <ocpp/v201/message_dispatcher.hpp>
+#include <ocpp/v201/message_handler.hpp>
+#include <ocpp/v201/messages/CancelReservation.hpp>
```
[the remark this answers is not in the document]
Comment (maaikez) [reply include/ocpp/v201/functional_blocks/reservation.hpp:6]: yep good idea, thanks!

File: include/ocpp/v201/functional_blocks/reservation.hpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+class EvseInterface;
+class EvseManagerInterface;
+
+class ReservationInterface : public MessageHandlerInterface {
+public:
+    virtual ~ReservationInterface(){};
+    virtual void on_reservation_status(const int32_t reservation_id, const ReservationUpdateStatusEnum status) = 0;
+    virtual ocpp::ReservationCheckStatus
+    is_evse_reserved_for_other(const EvseInterface& evse, const IdToken& id_token,
```
[the remark this answers is not in the document]
Comment (maaikez) [reply include/ocpp/v201/functional_blocks/reservation.hpp:22]: the `on_` functions already existed so I kept them like they were. The calling function of `is_evse_reserved_for_other` already has the `EvseInterface` of that evse so then it does not need to be searched for again.

File: lib/ocpp/v201/charge_point.cpp [hunk 4]
```diff
@@ -1281,10 +1289,12 @@ void ChargePoint::handle_message(const EnhancedMessage<v201::MessageType>& messa
             this->handle_heartbeat_response(json_message);
             break;
         case MessageType::ReserveNow:
-            this->handle_reserve_now_request(json_message);
-            break;
         case MessageType::CancelReservation:
-            this->handle_cancel_reservation_callback(json_message);
+            if (this->reservation != nullptr) {
+                this->reservation->handle_message(message);
```
[the remark this answers is not in the document]
Comment (maaikez) [reply lib/ocpp/v201/charge_point.cpp:1294]: I am only checking if reservation is a nullptr.

File: lib/ocpp/v201/functional_blocks/reservation.cpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        send_reserve_now_rejected_response(
+            call.uniqueId,
+            "No evse id was given while it should be sent in the request when NonEvseSpecific is disabled");
+        return;
+    }
+
+    const std::optional<int32_t> evse_id = request.evseId;
+
+    if (evse_id.has_value()) {
+        if (evse_id <= 0 || !evse_manager.does_evse_exist(evse_id.value())) {
```
[the remark this answers is not in the document]
Comment (maaikez) [reply lib/ocpp/v201/functional_blocks/reservation.cpp:113]: `does_evse_exist` still checks for >= 0. So I had to add this check.

File: lib/ocpp/v201/functional_blocks/reservation.cpp [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+void Reservation::send_reserve_now_rejected_response(const MessageId& unique_id, const std::string& status_info) {
+    ReserveNowResponse response;
+    response.status = ReserveNowStatusEnum::Rejected;
+    response.statusInfo = StatusInfo();
+    response.statusInfo->additionalInfo = status_info;
+    const ocpp::CallResult<ReserveNowResponse> call_result(response, unique_id);
+    this->message_dispatcher.dispatch_call_result(call_result);
+}
+
+bool Reservation::does_connector_exist(const uint32_t evse_id, std::optional<ConnectorEnum> connector_type) {
```
[the remark this answers is not in the document]
Comment (maaikez) [reply lib/ocpp/v201/functional_blocks/reservation.cpp:201]: well it calls the function from the evse class but first gets the evse from the evse manager which is needed there.

do you want the function in the evse manager then? that will then call does_connector_exist on the given evse?

File: tests/lib/ocpp/v201/functional_blocks/test_reservation.cpp (hunk 1 above)
Comment (maaikez) [reply tests/lib/ocpp/v201/functional_blocks/test_reservation.cpp:399]: Ok you are right. This is harder to maintain. I removed it. I don't see how I can test if the reservation was placed for a specific evse (and not for the other one). But we should have other test that do that.

File: tests/lib/ocpp/v201/functional_blocks/test_reservation.cpp [hunk 7]
```diff
@@ -0,0 +1,593 @@
...
+        EXPECT_EQ(response.status, ReserveNowStatusEnum::Rejected);
+        ASSERT_TRUE(response.statusInfo.has_value());
+        ASSERT_TRUE(response.statusInfo.value().additionalInfo.has_value());
+        EXPECT_EQ(response.statusInfo.value().additionalInfo.value(), "Reservation is not available");
+    }));
+
+    EvseMock& m1 = evse_manager.get_mock(1);
+    EvseMock& m2 = evse_manager.get_mock(2);
+
+    EXPECT_CALL(m1, get_connector_status(_)).Times(0);
```
Comment (marcemmers) [inline tests/lib/ocpp/v201/functional_blocks/test_reservation.cpp:195]: I think these are also not really relevant. Its already improved by quite a bit at least so would be ok to leave these for now.

File: include/ocpp/v201/functional_blocks/reservation.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcemmers) [reply include/ocpp/v201/functional_blocks/reservation.hpp:28]: Still looks a bit strange for the perspective of this class but I understand. Lets leave it like it is ;)

**Review by marcemmers — APPROVED** [review]

File: tests/lib/ocpp/v201/functional_blocks/test_reservation.cpp (hunk 7 above)
Comment (maaikez) [reply tests/lib/ocpp/v201/functional_blocks/test_reservation.cpp:195]: Yes i wanted to test if it is really doing nothing there. Only directly returns.

File: lib/ocpp/v201/functional_blocks/reservation.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (maaikez) [reply lib/ocpp/v201/functional_blocks/reservation.cpp:109]: https://github.com/EVerest/libocpp/issues/868
there is a bug report already

**Pull request merged**